### PR TITLE
Improved method for chessboard detection

### PR DESCRIPTION
**Pull request by mshabunin**
Ported from #4046 and #6851 by @HHI-VIT-IMC

>Modified and improved the method for chessboard detection. It is now faster and detects chessboards under difficult lighting condition as well as when the chessboard has strong out of plane rotations

Original PR has been transferred from 2.4 branch to master.

### CODE REVIEW

**Comment by HHI-VIT-IMC** [comment]
Am I right that now the improved chessboard detection is part of the next OpenCV release? If so, many thanks for the support.

**Comment by mshabunin** [comment]
@HHI-VIT-IMC , it will become part of new release (3.2) after this PR is merged.

Please, note that I've also reverted `min_aspect_ratio` and `max_aspect_ratio` in `icvGetQuadrangleHypotheses`, becasue otherwise the chessboard is detected on images without it.
I've also fixed the test which checked this behaviour (ported changes from #4187 and opencv/opencv_extra#261).

**Comment by mshabunin** [comment]
@vpisarev , could you please take a look?

**Comment by vpisarev** [comment]
@mshabunin, thanks for preparing this PR! Let's put it in so that users can test it thoroughly. Also, I would not mind if we put some extra test images with and without chessboards. :+1:

File: modules/calib3d/src/calibinit.cpp [hunk 1]
```diff
@@ -191,38 +195,204 @@ static void icvRemoveQuadFromGroup(CvCBQuad **quads, int count, CvCBQuad *q0);
...
-            CV_MAT_ELEM( *xy, float, i*2+1, 3 ) = CV_MAT_ELEM( *xy, float, i*2, 1 ) = pts2[j].y;
-            CV_MAT_ELEM( *xy, float, i*2, 2 ) = CV_MAT_ELEM( *xy, float, i*2, 3 ) = CV_MAT_ELEM( *xy, float, i*2, 5 ) = \
-                CV_MAT_ELEM( *xy, float, i*2+1, 0 ) = CV_MAT_ELEM( *xy, float, i*2+1, 1 ) = CV_MAT_ELEM( *xy, float, i*2+1, 4 ) = 0;
-            CV_MAT_ELEM( *xy, float, i*2, 4 ) = CV_MAT_ELEM( *xy, float, i*2+1, 5 ) = 1;
-            CV_MAT_ELEM( *uv, float, i*2, 0 ) = pts1[j].x;
-            CV_MAT_ELEM( *uv, float, i*2+1, 0 ) = pts1[j].y;
-            i++;
-        }
+      iIdx = i+ii;
+      if (iIdx > 0 && iIdx < 256)
```
Comment (alalek) [inline modules/calib3d/src/calibinit.cpp]:
```.diff
-iIdx > 0
+iIdx >= 0
```

#11591

File: modules/calib3d/src/calibinit.cpp [hunk 2]
```diff
@@ -191,38 +195,204 @@ static void icvRemoveQuadFromGroup(CvCBQuad **quads, int count, CvCBQuad *q0);
...
+/***************************************************************************************************/
+bool icvBinarizationHistogramBased( unsigned char* pucImg, int iCols, int iRows )
+{
+  int iMaxPix = iCols*iRows;
+  int iMaxPix1 = iMaxPix/100;
+  const int iNumBins = 256;
+  std::vector<int> piHistIntensity(iNumBins, 0);
+  std::vector<int> piHistSmooth(iNumBins, 0);
+  std::vector<int> piHistGrad(iNumBins, 0);
+  std::vector<int> piAccumSum(iNumBins, 0);
```
Comment (alalek) [inline modules/calib3d/src/calibinit.cpp]: `piAccumSum` is calculated, but not used anywhere.

File: modules/calib3d/src/calibinit.cpp [hunk 3]
```diff
@@ -191,38 +195,204 @@ static void icvRemoveQuadFromGroup(CvCBQuad **quads, int count, CvCBQuad *q0);
...
+  }
+  if ( iCntMaxima == 1)
+  {
+    iThresh = piMaxPos[0]/2;
+  }
+  else if ( iCntMaxima == 2)
+  {
+    iThresh = (piMaxPos[0] + piMaxPos[1])/2;
+  }
+  else // iCntMaxima >= 3
```
Comment (alalek) [inline modules/calib3d/src/calibinit.cpp]: Case `iCntMaxima == 0` is lost and incorrectly handled by "iCntMaxima >= 3" code.

#11590

File: modules/calib3d/src/calibinit.cpp [hunk 4]
```diff
@@ -191,38 +195,204 @@ static void icvRemoveQuadFromGroup(CvCBQuad **quads, int count, CvCBQuad *q0);
...
+    }
+
+    // CHECKING THRESHOLD FOR BLACK
+    int iMaxVal = piHistIntensity[piMaxPos[iIdxBGMax]];
+
+    //IF TOO CLOSE TO 255, jump to next maximum
+    if ( piMaxPos[iIdxBGMax] >= 250 && iIdxBGMax < iCntMaxima )
+    {
+      iIdxBGMax++;
+      iMaxVal = piHistIntensity[piMaxPos[iIdxBGMax]];
```
Comment (alalek) [inline modules/calib3d/src/calibinit.cpp]: Pre-condition above should avoid `piMaxPos` buffer overrun on this line:
```.diff
-iIdxBGMax < iCntMaxima
+iIdxBGMax + 1 < iCntMaxima
```

#11590

File: modules/calib3d/src/calibinit.cpp [hunk 5]
```diff
@@ -300,213 +464,258 @@ int cvFindChessboardCorners( const void* arr, CvSize pattern_size,
...
-                    // It splits chessboard corners
-                    //cvDilate( img, thresh_img, 0, 1 );
+      int max_quad_buf_size = 0;
 
-                    // empiric threshold level
-                    double mean = cvAvg( img ).val[0];
-                    int thresh_level = cvRound( mean - 10 );
-                    thresh_level = MAX( thresh_level, 10 );
+      //USE BINARY IMAGE COMPUTED USING icvBinarizationHistogramBased METHOD
+      cvDilate( thresh_img_new, thresh_img_new, 0, 1 );
```
Comment (alalek) [inline modules/calib3d/src/calibinit.cpp]: cvFindContours() (from icvGenerateQuads) modifies image, which becomes changed for next dilation iterations.

#11589

**Pull request merged**